include/unicode.h

```diff
@@ -0,0 +1,90 @@
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License along
+ *  with this program; if not, write to the Free Software Foundation, Inc.,
+ *  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+
+#ifndef DOSBOX_UNICODE_H
+#define DOSBOX_UNICODE_H
+
+#include <string>
+
+// Get recommended DOS code page to render the UTF-8 strings to. This
+// might not be the code page set using KEYB command, for example due
+// to emulated hardware limitations, or duplicated code page numbers
+uint16_t get_utf8_code_page();
+
+// Specifies what to do if the DOS code page does not contain character
+// representing given Unicode grapheme
+enum class UnicodeFallback {
+
+	// Convert all unknown graphemes to 0 - to be used in code like TREE
+	// command implementation, which has it's own specialized ASCII fallback
+	// drawing code
+	Null,
+
+	// Try to provide reasonable fallback using all the characters available
+	// in target DOS code page; use for features like clipboard content
+	// exchange with host system
+	Simple,
+
+	// Do not use certain DOS code page characters in order to draw boxes
+	// (tables) which are consistent; for example, if code page contains
+	// character '╠', but not '╣', both will be replaced with a fallback
+	// character ('║' for example)
+	Box
+};
+
+// Specifies how to interpret characters 0x00-0x1f and 0x7f in DOS strings
+enum class DosStringConvertMode {
+
+	// String contains control codes (new line, tabulation, delete, etc.)
+	WithControlCodes,
+
+	// String does not have any codes characters, consider all characters
+	// as screen codes
+	ScreenCodesOnly,
+
+	// String should not contain characters mentioned above
+	NoSpecialCharacters
+};
+
+// Convert the UTF-8 string to the format intended for display inside emulated
+// environment, or vice-versa. Code page '0' means a pure 7-bit ASCII. Functions
+// without 'code_page' parameters uses current DOS code page.
+std::string utf8_to_dos(const std::string& in_str,
+                        const DosStringConvertMode convert_mode,
+                        const UnicodeFallback fallback);
+std::string utf8_to_dos(const std::string& in_str,
+                        const DosStringConvertMode convert_mode,
+                        const UnicodeFallback fallback,
+                        const uint16_t code_page);
+std::string dos_to_utf8(const std::string& in_str,
+                        const DosStringConvertMode convert_mode);
+std::string dos_to_utf8(const std::string& in_str,
+                        const DosStringConvertMode convert_mode,
+                        const uint16_t code_page);
```

Comment:
minor: Please insert new lines between these to break up the block of text. It's quite unreadable as it is now.

Reply:
OK, changed.